File: src/services/firestoreService.ts
import {
  collection,
  doc,
  getDoc,
  setDoc,
  updateDoc,
  serverTimestamp,
  Timestamp,
  query,
  where,
  getDocs,
  addDoc,
  deleteDoc,
  orderBy,
} from 'firebase/firestore';
import { db } from '../config/firebase';
import { TravelerProfile, PartnerProfile, User, UserRole, Listing, ListingStatus, ListingCategory } from '../types';

// ========== User Service ==========

export const createUserDocument = async (
  uid: string,
  email: string,
  role: UserRole,
  emailVerified: boolean
): Promise<void> => {
  const userRef = doc(db, 'users', uid);
  await setDoc(userRef, {
    uid,
    email,
    role,
    emailVerified,
    createdAt: serverTimestamp(),
    updatedAt: serverTimestamp(),
  });
};

export const getUserDocument = async (uid: string): Promise<User | null> => {
  const userRef = doc(db, 'users', uid);
  const userSnap = await getDoc(userRef);

  if (userSnap.exists()) {
    const data = userSnap.data();
    return {
      ...data,
      createdAt: (data.createdAt as Timestamp).toDate(),
      updatedAt: (data.updatedAt as Timestamp).toDate(),
    } as User;
  }

  return null;
};

export const updateUserEmailVerification = async (uid: string): Promise<void> => {
  const userRef = doc(db, 'users', uid);
  await updateDoc(userRef, {
    emailVerified: true,
    updatedAt: serverTimestamp(),
  });
};

// ========== Traveler Profile Service ==========

export const createTravelerProfile = async (
  profile: Omit<TravelerProfile, 'createdAt' | 'updatedAt'>
): Promise<void> => {
  const travelerRef = doc(db, 'travelers', profile.userId);
  await setDoc(travelerRef, {
    ...profile,
    createdAt: serverTimestamp(),
    updatedAt: serverTimestamp(),
  });
};

export const getTravelerProfile = async (
  userId: string
): Promise<TravelerProfile | null> => {
  const travelerRef = doc(db, 'travelers', userId);
  const travelerSnap = await getDoc(travelerRef);

  if (travelerSnap.exists()) {
    const data = travelerSnap.data();
    return {
      ...data,
      createdAt: (data.createdAt as Timestamp).toDate(),
      updatedAt: (data.updatedAt as Timestamp).toDate(),
    } as TravelerProfile;
  }

  return null;
};

export const updateTravelerProfile = async (
  userId: string,
  updates: Partial<Omit<TravelerProfile, 'userId' | 'createdAt' | 'updatedAt'>>
): Promise<void> => {
  const travelerRef = doc(db, 'travelers', userId);
  await updateDoc(travelerRef, {
    ...updates,
    updatedAt: serverTimestamp(),
  });
};

// ========== Partner Profile Service ==========

export const createPartnerProfile = async (
  profile: Omit<PartnerProfile, 'createdAt' | 'updatedAt' | 'status'>
): Promise<void> => {
  const partnerRef = doc(db, 'partners', profile.userId);
  const profileData = {
    ...profile,
    status: 'pending' as const, // Partners need admin approval
    createdAt: serverTimestamp(),
    updatedAt: serverTimestamp(),
  };
  
  console.log('Creating partner profile:', profileData);
  await setDoc(partnerRef, profileData);
  console.log('Partner profile created successfully');
};

export const getPartnerProfile = async (
  userId: string
): Promise<PartnerProfile | null> => {
  const partnerRef = doc(db, 'partners', userId);
  const partnerSnap = await getDoc(partnerRef);

  if (partnerSnap.exists()) {
    const data = partnerSnap.data();
    return {
      ...data,
      createdAt: (data.createdAt as Timestamp).toDate(),
      updatedAt: (data.updatedAt as Timestamp).toDate(),
    } as PartnerProfile;
  }

  return null;
};

export const updatePartnerProfile = async (
  userId: string,
  updates: Partial<Omit<PartnerProfile, 'userId' | 'createdAt' | 'updatedAt' | 'status'>>
): Promise<void> => {
  const partnerRef = doc(db, 'partners', userId);
  await updateDoc(partnerRef, {
    ...updates,
    updatedAt: serverTimestamp(),
  });
};

// Admin function to approve partner
export const approvePartner = async (
  userId: string,
  adminId: string
): Promise<void> => {
  const partnerRef = doc(db, 'partners', userId);
  await updateDoc(partnerRef, {
    status: 'approved',
    approvedAt: serverTimestamp(),
    approvedBy: adminId,
    updatedAt: serverTimestamp(),
  });
};

// Admin function to reject partner
export const rejectPartner = async (
  userId: string,
  reason?: string
): Promise<void> => {
  const partnerRef = doc(db, 'partners', userId);
  await updateDoc(partnerRef, {
    status: 'rejected',
    rejectionReason: reason || '',
    updatedAt: serverTimestamp(),
  });
};

// Get all partners (admin only)
export const getAllPartners = async (): Promise<PartnerProfile[]> => {
  const partnersRef = collection(db, 'partners');
  const querySnapshot = await getDocs(partnersRef);

  return querySnapshot.docs.map((doc) => {
    const data = doc.data();
    return {
      userId: doc.id,
      ...data,
      createdAt: (data.createdAt as Timestamp).toDate(),
      updatedAt: (data.updatedAt as Timestamp).toDate(),
    } as PartnerProfile;
  });
};

// Get pending partners (admin only)
export const getPendingPartners = async (): Promise<PartnerProfile[]> => {
  const partnersRef = collection(db, 'partners');
  const q = query(partnersRef, where('status', '==', 'pending'));
  const querySnapshot = await getDocs(q);

  return querySnapshot.docs.map((doc) => {
    const data = doc.data();
    return {
      userId: doc.id,
      ...data,
      createdAt: (data.createdAt as Timestamp).toDate(),
      updatedAt: (data.updatedAt as Timestamp).toDate(),
    } as PartnerProfile;
  });
};

// ========== Listing Service ==========

export const createListing = async (
  listing: Omit<Listing, 'id' | 'createdAt' | 'updatedAt' | 'status'>
): Promise<string> => {
  const listingsRef = collection(db, 'listings');
  const listingData = {
    ...listing,
    status: 'pending' as ListingStatus, // Default status for new listings
    availability: {
      startDate: Timestamp.fromDate(listing.availability.startDate),
      endDate: Timestamp.fromDate(listing.availability.endDate),
    },
    createdAt: serverTimestamp(),
    updatedAt: serverTimestamp(),
  };
  
  const docRef = await addDoc(listingsRef, listingData);
  return docRef.id;
};

export const getListing = async (listingId: string): Promise<Listing | null> => {
  const listingRef = doc(db, 'listings', listingId);
  const listingSnap = await getDoc(listingRef);

  if (listingSnap.exists()) {
    const data = listingSnap.data();
    return {
      id: listingSnap.id,
      ...data,
      availability: {
        startDate: (data.availability.startDate as Timestamp).toDate(),
        endDate: (data.availability.endDate as Timestamp).toDate(),
      },
      createdAt: (data.createdAt as Timestamp).toDate(),
      updatedAt: (data.updatedAt as Timestamp).toDate(),
    } as Listing;
  }

  return null;
};

export const getPartnerListings = async (partnerId: string): Promise<Listing[]> => {
  const listingsRef = collection(db, 'listings');
  const q = query(listingsRef, where('partnerId', '==', partnerId), orderBy('createdAt', 'desc'));
  const querySnapshot = await getDocs(q);

  return querySnapshot.docs.map((doc) => {
    const data = doc.data();
    return {
      id: doc.id,
      ...data,
      availability: {
        startDate: (data.availability.startDate as Timestamp).toDate(),
        endDate: (data.availability.endDate as Timestamp).toDate(),
      },
      createdAt: (data.createdAt as Timestamp).toDate(),
      updatedAt: (data.updatedAt as Timestamp).toDate(),
    } as Listing;
  });
};

export const getApprovedListings = async (
  category?: ListingCategory
): Promise<Listing[]> => {
  const listingsRef = collection(db, 'listings');
  let q;
  
  if (category) {
    q = query(
      listingsRef,
      where('status', '==', 'approved'),
      where('category', '==', category),
      orderBy('createdAt', 'desc')
    );
  } else {
    q = query(
      listingsRef,
      where('status', '==', 'approved'),
      orderBy('createdAt', 'desc')
    );
  }

  const querySnapshot = await getDocs(q);

  return querySnapshot.docs.map((doc) => {
    const data = doc.data();
    return {
      id: doc.id,
      ...data,
      availability: {
        startDate: (data.availability.startDate as Timestamp).toDate(),
        endDate: (data.availability.endDate as Timestamp).toDate(),
      },
      createdAt: (data.createdAt as Timestamp).toDate(),
      updatedAt: (data.updatedAt as Timestamp).toDate(),
    } as Listing;
  });
};

export const updateListing = async (
  listingId: string,
  updates: Partial<Omit<Listing, 'id' | 'partnerId' | 'createdAt' | 'updatedAt'>>
): Promise<void> => {
  const listingRef = doc(db, 'listings', listingId);
  const updateData: any = {
    ...updates,
    updatedAt: serverTimestamp(),
  };

  // Convert dates if availability is being updated
  if (updates.availability) {
    updateData.availability = {
      startDate: Timestamp.fromDate(updates.availability.startDate),
      endDate: Timestamp.fromDate(updates.availability.endDate),
    };
  }

  await updateDoc(listingRef, updateData);
};

export const deleteListing = async (listingId: string): Promise<void> => {
  const listingRef = doc(db, 'listings', listingId);
  await deleteDoc(listingRef);
};

// Admin function to approve listing
export const approveListing = async (listingId: string): Promise<void> => {
  const listingRef = doc(db, 'listings', listingId);
  await updateDoc(listingRef, {
    status: 'approved',
    updatedAt: serverTimestamp(),
  });
};

// Admin function to reject listing
export const rejectListing = async (listingId: string): Promise<void> => {
  const listingRef = doc(db, 'listings', listingId);
  await updateDoc(listingRef, {
    status: 'rejected',
    updatedAt: serverTimestamp(),
  });
};

// Get all listings (admin only)
export const getAllListings = async (): Promise<Listing[]> => {
  const listingsRef = collection(db, 'listings');
  const q = query(listingsRef, orderBy('createdAt', 'desc'));
  const querySnapshot = await getDocs(q);

  return querySnapshot.docs.map((doc) => {
    const data = doc.data();
    return {
      id: doc.id,
      ...data,
      availability: {
        startDate: (data.availability.startDate as Timestamp).toDate(),
        endDate: (data.availability.endDate as Timestamp).toDate(),
      },
      createdAt: (data.createdAt as Timestamp).toDate(),
      updatedAt: (data.updatedAt as Timestamp).toDate(),
    } as Listing;
  });
};

// Get pending listings (admin only)
export const getPendingListings = async (): Promise<Listing[]> => {
  const listingsRef = collection(db, 'listings');
  const q = query(
    listingsRef,
    where('status', '==', 'pending'),
    orderBy('createdAt', 'desc')
  );
  const querySnapshot = await getDocs(q);

  return querySnapshot.docs.map((doc) => {
    const data = doc.data();
    return {
      id: doc.id,
      ...data,
      availability: {
        startDate: (data.availability.startDate as Timestamp).toDate(),
        endDate: (data.availability.endDate as Timestamp).toDate(),
      },
      createdAt: (data.createdAt as Timestamp).toDate(),
      updatedAt: (data.updatedAt as Timestamp).toDate(),
    } as Listing;
  });
};

// ========== Enhanced Search & Filter ==========

/**
 * Search listings with multiple filters
 */
export const searchListings = async (
  searchQuery?: string,
  category?: ListingCategory,
  minPrice?: number,
  maxPrice?: number,
  location?: string
): Promise<Listing[]> => {
  try {
    const listingsRef = collection(db, 'listings');
    let q = query(listingsRef, where('status', '==', 'approved'));

    // Add category filter
    if (category) {
      q = query(q, where('category', '==', category));
    }

    // Add price filters
    if (minPrice !== undefined) {
      q = query(q, where('price', '>=', minPrice));
    }
    if (maxPrice !== undefined) {
      q = query(q, where('price', '<=', maxPrice));
    }

    const querySnapshot = await getDocs(q);
    let listings = querySnapshot.docs.map((doc) => {
      const data = doc.data();
      return {
        id: doc.id,
        ...data,
        availability: {
          startDate: (data.availability.startDate as Timestamp).toDate(),
          endDate: (data.availability.endDate as Timestamp).toDate(),
        },
        createdAt: (data.createdAt as Timestamp).toDate(),
        updatedAt: (data.updatedAt as Timestamp).toDate(),
      } as Listing;
    });

    // Client-side filtering for search query and location
    if (searchQuery) {
      const query = searchQuery.toLowerCase();
      listings = listings.filter(
        listing =>
          listing.title.toLowerCase().includes(query) ||
          listing.description.toLowerCase().includes(query) ||
          listing.tags.some(tag => tag.toLowerCase().includes(query))
      );
    }

    if (location) {
      const loc = location.toLowerCase();
      listings = listings.filter(listing =>
        listing.location.toLowerCase().includes(loc)
      );
    }

    return listings;
  } catch (error) {
    console.error('Error searching listings:', error);
    throw error;
  }
};

/**
 * Get listings by category
 */
export const getListingsByCategory = async (
  category: ListingCategory
): Promise<Listing[]> => {
  try {
    const listingsRef = collection(db, 'listings');
    const q = query(
      listingsRef,
      where('status', '==', 'approved'),
      where('category', '==', category),
      orderBy('createdAt', 'desc')
    );

    const querySnapshot = await getDocs(q);
    return querySnapshot.docs.map((doc) => {
      const data = doc.data();
      return {
        id: doc.id,
        ...data,
        availability: {
          startDate: (data.availability.startDate as Timestamp).toDate(),
          endDate: (data.availability.endDate as Timestamp).toDate(),
        },
        createdAt: (data.createdAt as Timestamp).toDate(),
        updatedAt: (data.updatedAt as Timestamp).toDate(),
      } as Listing;
    });
  } catch (error) {
    console.error('Error getting listings by category:', error);
    throw error;
  }
};

/**
 * Get featured listings
 */
export const getFeaturedListings = async (): Promise<Listing[]> => {
  try {
    const listingsRef = collection(db, 'listings');
    const q = query(
      listingsRef,
      where('status', '==', 'approved'),
      orderBy('createdAt', 'desc')
    );

    const querySnapshot = await getDocs(q);
    return querySnapshot.docs.slice(0, 10).map((doc) => {
      const data = doc.data();
      return {
        id: doc.id,
        ...data,
        availability: {
          startDate: (data.availability.startDate as Timestamp).toDate(),
          endDate: (data.availability.endDate as Timestamp).toDate(),
        },
        createdAt: (data.createdAt as Timestamp).toDate(),
        updatedAt: (data.updatedAt as Timestamp).toDate(),
      } as Listing;
    });
  } catch (error) {
    console.error('Error getting featured listings:', error);
    return getApprovedListings();
  }
};

// ========== Booking Service ==========

/**
 * Create a new booking
 */
export const createBooking = async (
  bookingData: Omit<any, 'id' | 'createdAt' | 'updatedAt' | 'status' | 'paymentStatus'>
): Promise<string> => {
  try {
    const bookingsRef = collection(db, 'bookings');
    const docRef = await addDoc(bookingsRef, {
      ...bookingData,
      bookingDate: Timestamp.fromDate(bookingData.bookingDate),
      startDate: Timestamp.fromDate(bookingData.startDate),
      endDate: Timestamp.fromDate(bookingData.endDate),
      status: 'pending',
      paymentStatus: 'pending',
      createdAt: serverTimestamp(),
      updatedAt: serverTimestamp(),
    });
    return docRef.id;
  } catch (error) {
    console.error('Error creating booking:', error);
    throw error;
  }
};

/**
 * Get traveler's bookings
 */
export const getUserBookings = async (userId: string): Promise<any[]> => {
  try {
    const bookingsRef = collection(db, 'bookings');
    const q = query(
      bookingsRef,
      where('travelerId', '==', userId),
      orderBy('createdAt', 'desc')
    );

    const querySnapshot = await getDocs(q);
    return querySnapshot.docs.map((doc) => ({
      id: doc.id,
      ...doc.data(),
      bookingDate: (doc.data().bookingDate as Timestamp).toDate(),
      startDate: (doc.data().startDate as Timestamp).toDate(),
      endDate: (doc.data().endDate as Timestamp).toDate(),
      createdAt: (doc.data().createdAt as Timestamp).toDate(),
      updatedAt: (doc.data().updatedAt as Timestamp).toDate(),
    }));
  } catch (error) {
    console.error('Error getting user bookings:', error);
    throw error;
  }
};

/**
 * Get partner's bookings
 */
export const getPartnerBookings = async (partnerId: string): Promise<any[]> => {
  try {
    const bookingsRef = collection(db, 'bookings');
    const q = query(
      bookingsRef,
      where('partnerId', '==', partnerId),
      orderBy('createdAt', 'desc')
    );

    const querySnapshot = await getDocs(q);
    return querySnapshot.  docs.map((doc) => ({
      id: doc.id,
      ...doc.data(),
      bookingDate: (doc.data().bookingDate as Timestamp).toDate(),
      startDate: (doc.data().startDate as Timestamp).toDate(),
      endDate: (doc.data().endDate as Timestamp).toDate(),
      createdAt: (doc.data().createdAt as Timestamp).toDate(),
      updatedAt: (doc.data().updatedAt as Timestamp).toDate(),
    }));
  } catch (error) {
    console.error('Error getting partner bookings:', error);
    throw error;
  }
};

/**
 * Update booking status
 */
export const updateBookingStatus = async (
  bookingId: string,
  status: 'pending' | 'confirmed' | 'cancelled' | 'completed'
): Promise<void> => {
  try {
    const bookingRef = doc(db, 'bookings', bookingId);
    await updateDoc(bookingRef, {
      status,
      updatedAt: serverTimestamp(),
    });
  } catch (error) {
    console.error('Error updating booking status:', error);
    throw error;
  }
};

/**
 * Cancel booking
 */
export const cancelBooking = async (bookingId: string): Promise<void> => {
  try {
    await updateBookingStatus(bookingId, 'cancelled');
  } catch (error) {
    console.error('Error cancelling booking:', error);
    throw error;
  }
};

// ========== Favorites Service ==========

/**
 * Add listing to favorites
 */
export const addToFavorites = async (
  userId: string,
  listingId: string
): Promise<void> => {
  try {
    const favoriteRef = doc(db, 'favorites', `${userId}_${listingId}`);
    const listing = await getListing(listingId);
    
    if (listing) {
      await setDoc(favoriteRef, {
        userId,
        listingId,
        listingTitle: listing.title,
        listingImage: listing.images[0] || '',
        price: listing.price,
        createdAt: serverTimestamp(),
      });
    }
  } catch (error) {
    console.error('Error adding to favorites:', error);
    throw error;
  }
};

/**
 * Remove from favorites
 */
export const removeFromFavorites = async (
  userId: string,
  listingId: string
): Promise<void> => {
  try {
    const favoriteRef = doc(db, 'favorites', `${userId}_${listingId}`);
    await deleteDoc(favoriteRef);
  } catch (error) {
    console.error('Error removing from favorites:', error);
    throw error;
  }
};

/**
 * Get user's favorites
 */
export const getUserFavorites = async (userId: string): Promise<any[]> => {
  try {
    const favoritesRef = collection(db, 'favorites');
    const q = query(
      favoritesRef,
      where('userId', '==', userId),
      orderBy('createdAt', 'desc')
    );

    const querySnapshot = await getDocs(q);
    return querySnapshot.docs.map((doc) => ({
      ...doc.data(),
      createdAt: (doc.data().createdAt as Timestamp).toDate(),
    }));
  } catch (error) {
    console.error('Error getting user favorites:', error);
    throw error;
  }
};

/**
 * Check if listing is favorited
 */
export const isListingFavorited = async (
  userId: string,
  listingId: string
): Promise<boolean> => {
  try {
    const favoriteRef = doc(db, 'favorites', `${userId}_${listingId}`);
    const favoriteSnap = await getDoc(favoriteRef);
    return favoriteSnap.exists();
  } catch (error) {
    console.error('Error checking favorite status:', error);
    return false;
  }
};

/**
 * Get listing by ID (alias for getListing)
 */
export const getListingById = async (listingId: string): Promise<Listing | null> => {
  return getListing(listingId);
};

File: backend/services/firestore_service.py
"""
Firestore Service
Handles all Firestore database operations for the backend
"""

from config.firebase_admin import init_db
from typing import List, Dict, Any, Optional
from datetime import datetime

class FirestoreService:
    """Service for Firestore database operations"""
    
    def __init__(self):
        self.db = init_db()
    
    # Listings Operations
    async def get_all_listings(self, status: str = "approved") -> List[Dict[str, Any]]:
        """
        Get all listings with specified status
        """
        try:
            listings_ref = self.db.collection('listings')
            query = listings_ref.where('status', '==', status)
            docs = query.stream()
            
            listings = []
            for doc in docs:
                listing = doc.to_dict()
                listing['id'] = doc.id
                listings.append(listing)
            
            return listings
        except Exception as e:
            print(f"Error fetching listings: {e}")
            return []
    
    async def get_listing_by_id(self, listing_id: str) -> Optional[Dict[str, Any]]:
        """
        Get a single listing by ID
        """
        try:
            doc_ref = self.db.collection('listings').document(listing_id)
            doc = doc_ref.get()
            
            if doc.exists:
                listing = doc.to_dict()
                listing['id'] = doc.id
                return listing
            return None
        except Exception as e:
            print(f"Error fetching listing {listing_id}: {e}")
            return None
    
    async def search_listings(
        self, 
        category: Optional[str] = None,
        location: Optional[str] = None,
        min_price: Optional[float] = None,
        max_price: Optional[float] = None
    ) -> List[Dict[str, Any]]:
        """
        Search listings with filters
        """
        try:
            query = self.db.collection('listings').where('status', '==', 'approved')
            
            if category:
                query = query.where('category', '==', category)
            if location:
                query = query.where('location', '==', location)
            if min_price is not None:
                query = query.where('price', '>=', min_price)
            if max_price is not None:
                query = query.where('price', '<=', max_price)
            
            docs = query.stream()
            
            listings = []
            for doc in docs:
                listing = doc.to_dict()
                listing['id'] = doc.id
                listings.append(listing)
            
            return listings
        except Exception as e:
            print(f"Error searching listings: {e}")
            return []
    
    # User Operations
    async def get_user_profile(self, user_id: str, role: str) -> Optional[Dict[str, Any]]:
        """
        Get user profile (traveler or partner)
        """
        try:
            collection_name = 'travelers' if role == 'traveler' else 'partners'
            doc_ref = self.db.collection(collection_name).document(user_id)
            doc = doc_ref.get()
            
            if doc.exists:
                profile = doc.to_dict()
                profile['id'] = doc.id
                return profile
            return None
        except Exception as e:
            print(f"Error fetching user profile: {e}")
            return None
    
    async def get_traveler_profile(self, user_id: str) -> Optional[Dict[str, Any]]:
        """
        Get traveler profile by user ID
        """
        return await self.get_user_profile(user_id, 'traveler')
    
    async def get_partner_profile(self, user_id: str) -> Optional[Dict[str, Any]]:
        """
        Get partner profile by user ID
        """
        return await self.get_user_profile(user_id, 'partner')
    
    async def get_all_partners(self, status: str = "approved") -> List[Dict[str, Any]]:
        """
        Get all partners with specified status
        """
        try:
            partners_ref = self.db.collection('partners')
            query = partners_ref.where('status', '==', status)
            docs = query.stream()
            
            partners = []
            for doc in docs:
                partner = doc.to_dict()
                partner['id'] = doc.id
                partners.append(partner)
            
            return partners
        except Exception as e:
            print(f"Error fetching partners: {e}")
            return []
    
    # Booking Operations
    async def get_user_bookings(self, user_id: str, role: str = 'traveler') -> List[Dict[str, Any]]:
        """
        Get all bookings for a user (traveler or partner)
        """
        try:
            bookings_ref = self.db.collection('bookings')
            field_name = 'travelerId' if role == 'traveler' else 'partnerId'
            query = bookings_ref.where(field_name, '==', user_id)
            docs = query.stream()
            
            bookings = []
            for doc in docs:
                booking = doc.to_dict()
                booking['id'] = doc.id
                bookings.append(booking)
            
            return bookings
        except Exception as e:
            print(f"Error fetching bookings: {e}")
            return []
    
    async def get_booking_by_id(self, booking_id: str) -> Optional[Dict[str, Any]]:
        """
        Get a single booking by ID
        """
        try:
            doc_ref = self.db.collection('bookings').document(booking_id)
            doc = doc_ref.get()
            
            if doc.exists:
                booking = doc.to_dict()
                booking['id'] = doc.id
                return booking
            return None
        except Exception as e:
            print(f"Error fetching booking {booking_id}: {e}")
            return None
    
    # Partner-specific Operations
    async def get_partner_listings(self, partner_id: str) -> List[Dict[str, Any]]:
        """
        Get all listings for a specific partner
        """
        try:
            query = self.db.collection('listings').where('partnerId', '==', partner_id)
            docs = query.stream()
            
            listings = []
            for doc in docs:
                listing = doc.to_dict()
                listing['id'] = doc.id
                listings.append(listing)
            
            return listings
        except Exception as e:
            print(f"Error fetching partner listings: {e}")
            return []
    
    # Review Operations
    async def get_reviews(self, listing_id: Optional[str] = None, partner_id: Optional[str] = None) -> List[Dict[str, Any]]:
        """
        Get reviews for a listing or partner
        """
        try:
            query = self.db.collection('reviews')
            
            if listing_id:
                query = query.where('listingId', '==', listing_id)
            elif partner_id:
                query = query.where('partnerId', '==', partner_id)
            
            docs = query.stream()
            
            reviews = []
            for doc in docs:
                review = doc.to_dict()
                review['id'] = doc.id
                reviews.append(review)
            
            return reviews
        except Exception as e:
            print(f"Error fetching reviews: {e}")
            return []
    
    # Generic Operations
    async def query_collection(
        self, 
        collection: str, 
        filters: Optional[Dict[str, Any]] = None,
        limit: Optional[int] = None,
        order_by: Optional[str] = None
    ) -> List[Dict[str, Any]]:
        """
        Generic query method for any collection
        """
        try:
            query = self.db.collection(collection)
            
            if filters:
                for field, value in filters.items():
                    query = query.where(field, '==', value)
            
            if order_by:
                query = query.order_by(order_by)
            
            if limit:
                query = query.limit(limit)
            
            docs = query.stream()
            
            results = []
            for doc in docs:
                result = doc.to_dict()
                result['id'] = doc.id
                results.append(result)
            
            return results
        except Exception as e:
            print(f"Error querying collection {collection}: {e}")
            return []
    
    async def get_listings_since(self, since_date: datetime) -> List[Dict[str, Any]]:
        """
        Get listings updated since a specific date (for auto-sync)
        """
        try:
            query = self.db.collection('listings').where('updatedAt', '>=', since_date)
            docs = query.stream()
            
            listings = []
            for doc in docs:
                listing = doc.to_dict()
                listing['id'] = doc.id
                listings.append(listing)
            
            return listings
        except Exception as e:
            print(f"Error fetching listings since {since_date}: {e}")
            return []
    
    async def create_document(self, collection: str, data: Dict[str, Any], document_id: Optional[str] = None) -> Optional[str]:
        """
        Create a new document in a collection
        """
        try:
            if document_id:
                doc_ref = self.db.collection(collection).document(document_id)
                doc_ref.set(data)
                return document_id
            else:
                doc_ref = self.db.collection(collection).add(data)
                return doc_ref[1].id
        except Exception as e:
            print(f"Error creating document in {collection}: {e}")
            return None
    
    async def update_document(self, collection: str, document_id: str, data: Dict[str, Any]) -> bool:
        """
        Update an existing document
        """
        try:
            doc_ref = self.db.collection(collection).document(document_id)
            doc_ref.update(data)
            return True
        except Exception as e:
            print(f"Error updating document {document_id} in {collection}: {e}")
            return False
    
    async def delete_document(self, collection: str, document_id: str) -> bool:
        """
        Delete a document from a collection
        """
        try:
            doc_ref = self.db.collection(collection).document(document_id)
            doc_ref.delete()
            return True
        except Exception as e:
            print(f"Error deleting document {document_id} from {collection}: {e}")
            return False

# Singleton instance
firestore_service = FirestoreService()
